### PR TITLE
Phase 1 of ISet support

### PR DESCRIPTION
**Pull request by nirinchev**
## Description

This is part 1 of https://github.com/realm/realm-dotnet/issues/2099. It stubs the .NET implementation and adds weaver support. It also implements some the Set methods for unmanaged collections, but is not wired up to the native set.

Fixes https://github.com/realm/realm-dotnet/issues/2064.
Fixes https://github.com/realm/realm-dotnet/issues/2089.

### CODE REVIEW

File: Realm/Realm.Fody/ModuleWeaver.cs [hunk 1]
```diff
@@ -469,33 +470,51 @@ private WeaveResult WeaveProperty(PropertyDefinition prop, TypeDefinition type,
...
+                                        new GenericInstanceMethod(_references.RealmObject_GetListValue) { GenericArguments = { elementType } },
+                                        concreteListConstructor);
+                    break;
+                case RealmCollectionType.ISet:
+                    var concreteSetConstructor = _references.System_Collections_Generic_HashSetOfT_Constructor.MakeHostInstanceGeneric(elementType);
 
-            ReplaceListGetter(prop, backingField, columnName,
-                              new GenericInstanceMethod(_references.RealmObject_GetListValue) { GenericArguments = { elementType } },
-                              concreteListConstructor);
+                    // weaves set getter which also sets backing to List<T>, forcing it to accept us setting it post-init
```
Comment (LaPeste) [inline Realm/Realm.Fody/ModuleWeaver.cs:507]: Hashset<T> should be the backing here.

File: Tests/Weaver/AssemblyToProcess/ClassWithStaticProperty.cs [hunk 2]
```diff
@@ -26,12 +27,13 @@ public class Dog : RealmObject
...
         public string Name { get; set; }
 
-        private static Dog myDog;
+        private static readonly Dog myDog = new Dog();
 
         public static Dog MyOnlyDog
         {
             get
             {
+                Console.WriteLine("");
```
Comment (papafe) [inline Tests/Weaver/AssemblyToProcess/ClassWithStaticProperty.cs:36]: Why was this needed?

Comment (nirinchev) [reply Tests/Weaver/AssemblyToProcess/ClassWithStaticProperty.cs:36]: Apf... this was a silly way to silence a VS message that this should be an auto property. I'll change it to be more meaningful.

File: Tests/Weaver/Realm.FakeForWeaverTests/RealmSet.cs [hunk 3]
```diff
@@ -0,0 +1,121 @@
+﻿////////////////////////////////////////////////////////////////////////////
+//
+// Copyright 2016 Realm Inc.
```
Comment (nirinchev) [inline Tests/Weaver/Realm.FakeForWeaverTests/RealmSet.cs:3]: Fix

File: Realm/Realm/Dynamic/MetaRealmSet.cs [hunk 4]
```diff
@@ -0,0 +1,50 @@
+﻿////////////////////////////////////////////////////////////////////////////
+//
+// Copyright 2016 Realm Inc.
```
Comment (nirinchev) [inline Realm/Realm/Dynamic/MetaRealmSet.cs:3]: fix all dates 😑
